### PR TITLE
Neon: Do not put functions into .rodata section

### PR DESCRIPTION
**Pull request by rui314**
jsimd_neon.S contains `.section .rodata` directive at the beginning
of the file but didn't have `.text` before function definitions.
As a result, executable code in this file was written to .rodata,
which makes the entire .rodata section in libjpeg.so executable.
This is bad from a security standpoint.

### CODE REVIEW

**Comment by dcommander** [comment]
To the best of my understanding, this PR would reverse b542e4c8e932679273f9ce16c2b5f57104a9111f (refer to #318), which moved constants into `.rodata` in order to support execute-only memory.  @jwright-arm can comment further, but since the Neon GAS code is only used by default with legacy compilers (libjpeg-turbo 2.1 introduced a full intrinsics implementation of the Neon SIMD extensions, and the intrinsics implementation is used by default when building with Clang), this PR seems minimally unnecessary and perhaps even harmful.

**Comment by rui314** [comment]
The constants are above the `.text` line, so this does not reverse what https://github.com/libjpeg-turbo/libjpeg-turbo/commit/b542e4c8e932679273f9ce16c2b5f57104a9111f did.

Please take a look at https://github.com/libjpeg-turbo/libjpeg-turbo/commit/240ba417aa4b3174850d05ea0d22dbe5f80553c1. That is the change that (presumably accidentally) removed the `.text` line from the assembly file, making everything in this file including functions to be put into .rodata.

**Comment by rui314** [comment]
Here is the background: I'm the original author of the LLVM linker, lld, and I'm creating a new one called "mold" right now (https://github.com/rui314/mold). I'm currently porting mold to AArch64 and building bunch of programs for testing. I noticed that libjpeg-turbo failed to pass its unit tests when linked with my linker because it jumps into .rodata. I thought that it was my linker's bug, but it turned out that GNU ld was automagically and silently making .rodata into an executable segment when a source .rodata contains functions. jsimd_neon.S.o indeed contained functions in .rodata. I believe it is not intentional, so I made this patch.

**Comment by dcommander** [comment]
OK, thanks.  In the future, you will get a better response if you specify that the issue is a regression.  I will investigate.

**Comment by rui314** [comment]
I'm not sure why this file was chosen to be used if newer and better one is available. I just built it on the most recent version of Arch Linux with the default settings. It could be an Arch Linux's issue, but I'm not sure.

**Comment by dcommander** [comment]
@rui314 If the libjpeg-turbo build system detects that the compiler has a full Neon intrinsics implementation, then the build system will set the default value of the `NEON_INTRINSICS` CMake variable to `1`.  Otherwise, the build system will set the default value of the `NEON_INTRINSICS` CMake variable to `0`.  If `NEON_INTRINSICS` is `1`, then libjpeg-turbo will use Neon intrinsics for all SIMD algorithms.  If `NEON_INTRINSICS` is `0`, then libjpeg-turbo will continue to use the older GAS implementation for any SIMD algorithms whose intrinsics implementations would not perform well with the compiler because of the compiler's limited support for Neon intrinsics.

In short, when using GCC, you must (for the moment) continue to use the GAS implementation of certain libjpeg-turbo SIMD algorithms in order to avoid a serious performance regression relative to libjpeg-turbo 2.0.x.

I will push a fix shortly.

**Comment by dcommander** [comment]
@jwright-arm You assert that newer versions of GCC shouldn't suffer from the same performance regression, but the build system assumes that the performance regression will exist if `vld1_s16_x3()`, `vld1_u16_x2()`, or `vld1q_u8_x4()` doesn't exist.  That seems to be a valid assumption for all GCC releases I've tested, up to and including 10.3.x.  Do you observe otherwise?
